ssd1306/ssd1306.c

```diff
@@ -189,12 +176,26 @@ void ssd1306_UpdateScreen(void) {
     //  * 32px   ==  4 pages
     //  * 64px   ==  8 pages
     //  * 128px  ==  16 pages
-    for(uint8_t i = 0; i < SSD1306_HEIGHT/8; i++) {
-        ssd1306_WriteCommand(0xB0 + i); // Set the current RAM page address.
-        ssd1306_WriteCommand(0x00);
-        ssd1306_WriteCommand(0x10);
-        ssd1306_WriteData(&SSD1306_Buffer[SSD1306_WIDTH*i],SSD1306_WIDTH);
-    }
+#if defined(SSD1306_USE_SPI)
+	ssd1306_WriteCommand(0xB0); // Set the current RAM page address.
+    ssd1306_WriteCommand(0x00);
+    ssd1306_WriteCommand(0x10);
+	HAL_GPIO_WritePin(SSD1306_CS_Port, SSD1306_CS_Pin, GPIO_PIN_RESET); // select OLED
+    HAL_GPIO_WritePin(SSD1306_DC_Port, SSD1306_DC_Pin, GPIO_PIN_SET); // data
+#if defined(SSD1306_USE_DMA)
+    HAL_SPI_Transmit_DMA(&SSD1306_SPI_PORT, SSD1306_Buffer, SSD1306_BUFFER_SIZE);
+#else
+    HAL_SPI_Transmit(&SSD1306_SPI_PORT, SSD1306_Buffer, SSD1306_BUFFER_SIZE, HAL_MAX_DELAY);
+#endif
+    HAL_GPIO_WritePin(SSD1306_CS_Port, SSD1306_CS_Pin, GPIO_PIN_SET); // un-select OLED
+
+#elif defined(SSD1306_USE_I2C)
+#if defined(SSD1306_USE_DMA)
+    HAL_I2C_Mem_Write_DMA(&SSD1306_I2C_PORT, SSD1306_I2C_ADDR, 0x40, 1, SSD1306_Buffer, SSD1306_BUFFER_SIZE);
+#else
+    HAL_I2C_Mem_Write(&SSD1306_I2C_PORT, SSD1306_I2C_ADDR, 0x40, 1, SSD1306_Buffer, SSD1306_BUFFER_SIZE, HAL_MAX_DELAY);
+#endif
+#endif
```

Comment:
Wouldn't it be better to encapsulate these ifdefs into ssd1306_WriteData?

Reply:
Yes, these macros can be placed in ssd1306_WriteData, which seems better, you're right. I'll do this.
Data is transferred. I tested my code on my STM32F103C8T6 under I2C with DMA, I2C without DMA, SPI with DMA and SPI without DMA with function ssd1306_TestFPS(). FPS is higher when DMA is enabled.

Reply:
Unfortunately, DMA and non-DMA versions of SPI and I2C transmit functions are not interchangeable.

HAL_SPI_Transmit_DMA does not wait for the previous DMA transaction to complete. Therefore, if you call HAL_SPI_Transmit_DMA during an active SPI DMA transaction, it will return HAL_BUSY and won't send any data. Since this situation is not handled in the code, ssd1306_WriteData will occasionally lose data.

(see: [HAL_SPI_Transmit_DMA source code](https://github.com/STMicroelectronics/STM32CubeF1/blob/c750eab6990cac35ab05020793b0221ecc1a8ce5/Drivers/STM32F1xx_HAL_Driver/Src/stm32f1xx_hal_spi.c#L1674))

HAL_I2C_Mem_Write_DMA has a different but equally dangerous pitfall. The function tries to wait until I2C is ready for the next DMA transfer. It does this by checking I2C_FLAG_BUSY in a loop. As soon as I2C becomes free, it starts a new DMA transfer. I2C_FLAG_BUSY is cleared in I2C_DMAXferCplt, which in turn is called by the DMA interrupt handler.

When you call HAL_I2C_Mem_Write_DMA from an interrupt handler, and the Interrupt Priority Group of the DMA interrupt is equal or less than the current interrupt priority group, you have a deadlock. The DMA interrupt will not be able to reset I2C_FLAG_BUSY because it cannot interrupt the interrupt handler being currently executed. This can be resolved by configuring proper interrupt priority grouping, but the PR does not have anything related to such configuration. This deadlock condition will freeze the firmware for HAL_MAX_DELAY ms, which practically means forever.

(see: [HAL_I2C_Mem_Write_DMA](https://github.com/STMicroelectronics/STM32CubeF1/blob/c750eab6990cac35ab05020793b0221ecc1a8ce5/Drivers/STM32F1xx_HAL_Driver/Src/stm32f1xx_hal_i2c.c#L3142))

Although this PR offers some degree of performance improvement, it also introduces the issues described above, which degrade stability.

Reply:
Really thanks for your detailed explianation! I'll fix. Learned a lot from you!

Reply:
In fact, neither `HAL_I2C_Mem_Write_DMA` nor `HAL_I2C_Mem_Write` waits for the previous transfer to be done.
[`HAL_I2C_Mem_Write_DMA` returned if `hi2cx->State != HAL_I2C_STATE_READY`.](https://github.com/STMicroelectronics/STM32CubeF1/blob/c750eab6990cac35ab05020793b0221ecc1a8ce5/Drivers/STM32F1xx_HAL_Driver/Src/stm32f1xx_hal_i2c.c#L3303)
[`HAL_I2C_Mem_Write` returned if `hi2cx->State != HAL_I2C_STATE_READY`.](https://github.com/STMicroelectronics/STM32CubeF1/blob/c750eab6990cac35ab05020793b0221ecc1a8ce5/Drivers/STM32F1xx_HAL_Driver/Src/stm32f1xx_hal_i2c.c#L2645)
Ah, I'd better manually use `while( hi2cx.State != HAL_I2C_STATE_READY )` to wait.